src/link/Dwarf.zig

```diff
@@ -31,27 +31,89 @@ dbg_line_fn_free_list: std.AutoHashMapUnmanaged(*SrcFn, void) = .{},
 dbg_line_fn_first: ?*SrcFn = null,
 dbg_line_fn_last: ?*SrcFn = null,
 
-/// A list of `TextBlock` whose corresponding .debug_info tags have surplus capacity. /// This is the same concept as `text_block_free_list`; see those doc comments.
-dbg_info_decl_free_list: std.AutoHashMapUnmanaged(*DebugInfoAtom, void) = .{},
-dbg_info_decl_first: ?*DebugInfoAtom = null,
-dbg_info_decl_last: ?*DebugInfoAtom = null,
+/// A list of `Atom`s whose corresponding .debug_info tags have surplus capacity.
+/// This is the same concept as `text_block_free_list`; see those doc comments.
+atom_free_list: std.AutoHashMapUnmanaged(*Atom, void) = .{},
+atom_first: ?*Atom = null,
+atom_last: ?*Atom = null,
 
 abbrev_table_offset: ?u64 = null,
 
+/// TODO replace with InternArena
 /// Table of debug symbol names.
 strtab: std.ArrayListUnmanaged(u8) = .{},
 
-pub const DebugInfoAtom = struct {
+/// Lives only as long as the analysed Decl.
+/// Allocated with `initDeclState`.
+/// Freed with `commitDeclState`.
+decl_state: ?DeclState = null,
```

Comment:
I think it would be better to eliminate this field. It does not belong here; it belongs as ephemeral data in the space between init() and commit(). init() should return some kind of work-in-progress type, and commit() should consume it.

It took me a long time to understand what the purpose of this field is based on its existence inside this struct because it does not belong here; the Dwarf struct is global to the whole compilation but this `decl_state` is for exactly one Decl.

Furthermore, it encourages buggy code, such as the problem I'm currently running into, where one of the `try` has triggered in Elf.zig between init() and commit(), causing the `decl_state` to be non-null and the assertion tripping next time init() is called.

I suggest to remove this field, and see where that naturally leads the code to be instead.

Reply:
e39c86399da565d3855d54b242bf7b5940e3924d

Reply:
Thanks for pointing this out @andrewrk - fix landed in https://github.com/ziglang/zig/pull/11349